### PR TITLE
[FBX - WIP] Various improvements for blender FBX specifically.

### PR DESCRIPTION
**Pull request by RevoluPowered**
Added a sanity check test, because I saw a bug on my PC which seems very rare, can be removed when this pr is ready I don't mind.

Fixes:
- normal data is now working from blender fbx files
- skinned meshes work more of the time from blender.
- blender fbx's *only* will reparent the mesh to the skeleton.
- blender has no support for pivots in an fbx.
- kenny character model skin is fixed.

:) 🍰 
Screenshot 2021-02-28 at 18 33 26

TODO:
- Fix animation time basis
- Remove pivots for blender fbx keyframes 
- default keys for blender fbx tracks probably need disabled

Bugs in engine:
- aabb's of skinned meshes are of the mesh without the skin (orientation then for the aabb is wrong, but maybe we can fix this?)

### CODE REVIEW

**Comment by fire** [comment]
If you can provide a sample project, my test plan is to run the project and do a visual inspection and then approve the pr.

File: tests/test_compiler_sanity.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef TEST_SANITY_H
+#define TEST_SANITY_H
+
+#include "core/os/os.h"
+#include "core/string/ustring.h"
+#include "scene/resources/material.h"
```
Comment (akien-mga) [inline tests/test_compiler_sanity.h:36]: Those headers don't seem used (at least `material.h`, didn't check that it compiles without though).

But I'd suggest moving this to a different PR as it's not related to FBX import (and that PR might need more details on what this test is for and how to reproduce the compiler issue it's trying to detect).

Comment (RevoluPowered) [reply tests/test_compiler_sanity.h:36]: ok, it was more for my own education as I had some bugs locally and some weird git behaviour gave me outdated modifications to files for some bizarre reason :)

I'll remove this

**Comment by RevoluPowered** [comment]
> If you can provide a sample project, my test plan is to run the project and do a visual inspection and then approve the pr.

I'll see if I can provide test files, but it really depends when I can.

I'm using kenny's paid for variant to test the skin fixes specifically.

I'll be working on keyframes fixes for blender next

**Comment by RevoluPowered** [comment]
- 🍰 keyframes are fixed.
- it wont sample too many keyframes if there is no movement
- time basis is fixed
- i had to remove from to rotation for blender to work
- added flag to set the track to be an imported non-editable track

todo:
- fix compat bug with skin formula for maya fbx

File: modules/fbx/data/fbx_skeleton.cpp [hunk 2]
```diff
@@ -69,7 +69,7 @@ void FBXSkeleton::init_skeleton(const ImportState &state) {
 			// Make sure the bone name is unique.
 			const String bone_name = bone->bone_name;
 			int same_name_count = 0;
-			for (int y = x; y < skeleton_bone_count; y++) {
+			for (int y = x + 1; y < skeleton_bone_count; y++) {
```
Comment (akien-mga) [inline modules/fbx/data/fbx_skeleton.cpp:72]: This fixes #43820 right?

Comment (RevoluPowered) [reply modules/fbx/data/fbx_skeleton.cpp:72]: yep

Comment (BastiaanOlij) [reply modules/fbx/data/fbx_skeleton.cpp:72]: yeah we already merged that in 3 I think

Comment (akien-mga) [reply modules/fbx/data/fbx_skeleton.cpp:72]: I extracted it in its own PR #47458 as this one is still WIP and will likely not make it in time for 3.3-stable.

Comment (akien-mga) [reply modules/fbx/data/fbx_skeleton.cpp:72]: Ah you were right @BastiaanOlij, this has already been fixed in `3.x` via 9c296bcb59c. Should really have gone to `master` and be cherry-picked instead :)

Comment (BastiaanOlij) [reply modules/fbx/data/fbx_skeleton.cpp:72]: yeah when I did that one Gordon said he was already including it in one of his fixes so :)

**Comment by akien-mga** [comment]
Superseded by #59691 and #59653.
